django/test/runner.py

```diff
@@ -602,7 +603,12 @@ def run_tests(self, test_labels, extra_tests=None, **kwargs):
         self.setup_test_environment()
         suite = self.build_suite(test_labels, extra_tests)
         old_config = self.setup_databases()
-        self.run_checks()
+        try:
+            self.run_checks()
+        except SystemCheckError:
+            self.teardown_databases(old_config)
+            self.teardown_test_environment()
+            raise
         result = self.run_suite(suite)
         self.teardown_databases(old_config)
         self.teardown_test_environment()
```

Comment:
You current approach will silence any `SystemCheckError` raised by `run_checks` if `teardown_databases` and `teardown_test_environment` fail.

What about

```python
run_failed = False
try:
    self.run_checks()
    result = self.run_suite(suite)
except Exception:
    run_failed = True
    raise
finally:
    try:
        self.teardown_databases(old_config)
        self.teardown_test_environment()
    except Exception:
        if not run_failed:
           raise
return self.suite_result(suite, result)
```